pyomo/contrib/trustregion/TRF.py

```diff
@@ -122,128 +102,312 @@ def TRF(m, eflist, config):
+        default = 0.05,
+        domain=In(NumericRange(0, 1, 0, (False, False))),
+        description="Lower ratio test parameter (eta_1). "
+                    "Must satisfy: 0 < eta_1 <= eta_2 < 1. "
+                    "Default = 0.05."
+    ))
+    CONFIG.declare('ratio test param eta_2', ConfigValue(
+        default = 0.2,
+        domain=In(NumericRange(0, 1, 0, (False, False))),
+        description="Lower ratio test parameter (eta_2). "
+                    "Must satisfy: 0 < eta_1 <= eta_2 < 1. "
+                    "Default = 0.2."
+    ))
+    ### Filter
+    CONFIG.declare('maximum feasibility', ConfigValue(
+        default=50.0,
+        domain=PositiveFloat,
+        description="Maximum allowable feasibility measure (theta_max). "
+                    "Parameter for use in filter method."
+                    "Default = 50.0."
+    ))
+    CONFIG.declare('filter param gamma theta', ConfigValue(
+        default=0.01,
+        domain=In(NumericRange(0, 1, 0, (False, False))),
+        description="Fixed filter parameter (gamma_theta) within (0, 1). "
+                    "Default = 0.01"
+    ))
+    CONFIG.declare('filter param gamma f', ConfigValue(
+        default=0.01,
+        domain=In(NumericRange(0, 1, 0, (False, False))),
+        description="Fixed filter parameter (gamma_f) within (0, 1). "
+                    "Default = 0.01"
+    ))
+
+    return CONFIG
+
+
+@SolverFactory.register(
+    'trustregion',
+    doc='Trust region algorithm "solver" for black box/glass box optimization')
+class TrustRegionSolver(object):
+    """
+    The Trust Region Solver is a 'solver' based on the 2016/2018/2020 AiChE
+    papers by Eason (2016/2018), Yoshio (2020), and Biegler.
 
-                config.trust_radius = min(
-                    max(config.gamma_e*stepNorm, config.trust_radius),
-                    config.radius_max)
+    """
 
-            else:
-                logger.iterlog.thetaStep = True
-
-                fe = FilterElement(
-                    obj - config.gamma_f*theta,
-                    (1 - config.gamma_theta)*theta)
-                filteR.addToFilter(fe)
-
-                # Calculate rho for theta step trust region update
-                rhok = 1 - ((theta - config.ep_i) /
-                            max(thetak, config.ep_i))
-                if rhok < config.eta1:
-                    config.trust_radius = max(config.gamma_c*stepNorm,
-                                      config.delta_min)
-                elif rhok >= config.eta2:
-                    config.trust_radius = min(
-                        max(config.gamma_e*stepNorm, config.trust_radius),
-                        config.radius_max)
-
-        # Accept step
-        rebuildROM = True
-        xk, yk, zk = cloneXYZ(x, y, z)
-        thetak = theta
-        objk = obj
-
-
-    logger.printVectors()
-#    problem.reverseTransform()
+    def __init__(self, **kwds):
+        self._CONFIG = _trf_config()
```

Comment:
The general pattern for ConfigDict is to create a definitive master CONFIG (as a class attribute) and then creating an instance-specific copy from it; that is:
```python
class TrustRegionSolver(object):
    CONFIG = _trf_config()

    def __init__(self, **kwds):
        self.config = self.CONFIG(kwds)

   def solve(**kwds):
        config = self.config(kwds.pop('options', {})
        config.set_value(kwds)
        return trust_region_method(
                model,
                degrees_of_freedom_variables,
                ext_fcn_surrogate_map_rule,
                config)
```
This allows for 3 levels of config options: `TrustRegionSOlver.CONFIG` are persistent and shared across all future instances of the solver, `self.config`, which is specific to the current instance, and `config`, which is completely ephemeral (unique to that `solve()`).